### PR TITLE
#950: refactor options line parser v2

### PR DESCRIPTION
**Pull request by tomuben**

### CODE REVIEW

File: exaudfclient/base/javacontainer/script_options/converter.cc [hunk 1]
```diff
@@ -1,123 +1,56 @@
...
-            if (delim != std::string::npos) {
-                m_jvmOptions.push_back(jvmOption.substr(start, delim - start));
-            }
-            else {
-                m_jvmOptions.push_back(jvmOption.substr(start));
-                break;
-            }
-        }
+void Converter::convertScriptClassName(const std::string & value) {
+    DBGVAR(std::cerr, value);
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/converter.cc:32]: Is this a left over?

File: exaudfclient/base/javacontainer/script_options/converter.cc [hunk 2]
```diff
@@ -1,123 +1,56 @@
...
-    // package definition). Otherwise we don't recognize if the script imports its self
-    m_importedScriptChecksums.insert(scriptToMd5(src_scriptCode.c_str()));
-    while (true) {
-        std::pair<std::string, size_t> nextImportStatement;
-        m_scriptOptionsParser->getNextImportScript(src_scriptCode, nextImportStatement,
-                                                     [&](const std::string& msg){m_throwException("F-UDF-CL-SL-JAVA-1604" + msg);});
-        if (nextImportStatement.first == "")
+void Converter::convertJvmOption(const std::string & value) {
+    for (size_t start = 0, delim = 0; ; start = delim + 1) {
+        start = value.find_first_not_of(m_whitespace, start);
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/converter.cc:40]: Here, we need to later see, how we deal with white spaces in the jvm option

File: exaudfclient/base/javacontainer/script_options/converter.h [hunk 3]
```diff
@@ -15,33 +15,32 @@ namespace SWIGVMContainers {
...
-        void getExternalJarPaths(std::string & src_scriptCode);
+        void convertExternalJar(const std::string & value);
+
+        void convertScriptClassName(const std::string & value);
 
-        void getScriptClassName(std::string & src_scriptCode);
+        void convertJvmOption(const std::string & value);
 
-        void convertImportScripts(std::string & src_scriptCode);
+        const std::set<std::string> & getJarPaths() const {
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/converter.h:29]: Hm, you use a move for the Jvm options, but not here, is there a reason.

File: exaudfclient/base/javacontainer/script_options/converter.h [hunk 4]
```diff
@@ -15,33 +15,32 @@ namespace SWIGVMContainers {
...
     private:
-    
-        std::unique_ptr<ScriptOptionsParser> m_scriptOptionsParser;
-        
-        std::function<void(const std::string &msg)> m_throwException;
 
-        std::vector<std::string>& m_jvmOptions;
+        std::vector<std::string> m_jvmOptions;
         
         std::set<std::string> m_jarPaths;
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/converter.h:41]: Set for jarPaths might change the order of jars. The order of jars might define which classes get loaded, when multiple jars contain classes with the same full name. Did we use a set before?

File: exaudfclient/base/javacontainer/script_options/converter.cc (hunk 1 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/converter.cc:32]: Yes, it is. Removed it.

File: exaudfclient/base/javacontainer/script_options/converter.h (hunk 3 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/converter.h:29]: The JavaVMImpl class iterates over all items of this set in the constructor, anyway. So there is no need to move the set. On the other hand, if you say that we should have a common interface for the return values, maybe then it makes sense...

File: exaudfclient/base/javacontainer/script_options/parser_legacy.cc [hunk 5]
```diff
@@ -9,78 +12,181 @@ namespace JavaScriptOptions {
...
+                    throwException("F-UDF-CL-SL-JAVA-1615: Failure while importing scripts");
+            }
+            const char *importScriptCode = metaData->moduleContent(newScript.c_str());
+            const char *exception = metaData->checkException();
+            if (exception)
+                throwException("F-UDF-CL-SL-JAVA-1616: " + std::string(exception));
+            if (importedScriptChecksums.addScript(importScriptCode)) {
+                // Script has not been imported yet
+                // If this imported script contains %import statements
+                // they will be resolved in the recursion.
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_legacy.cc:118]: Not directly a recursion

File: exaudfclient/base/javacontainer/script_options/converter.h (hunk 4 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/converter.h:41]: Yes, it was a set before see https://github.com/exasol/script-languages/blob/4ebbbac7f7c0b72ee268f1aa59effb0a3d44c81b/exaudfclient/base/javacontainer/javacontainer_impl.h#L53

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 6]
```diff
@@ -21,94 +21,94 @@ TEST(JavaContainer, basic_jar) {
...
-     * Note: The option "DEF" is wrong and causes UDF's to crash!
-     *       The correct option would be '-Dhttp.agent=\"ABC DEF\"'
-     */
-    const std::vector<std::string> expectedJVMOptions = {   "-Dhttp.agent=\"ABC", "DEF\"", "-Xms128m", "-Xmx128m", "-Xss512k",
-                                                            "-XX:ErrorFile=/tmp/hs_err_pid%p.log",
-                                                            "-Djava.class.path=/tmp:/exaudf/base/javacontainer/exaudf_deploy.jar",
-                                                            "-XX:+UseSerialGC" };
-    EXPECT_EQ(vm.getJavaVMInternalStatus().m_jvmOptions, expectedJVMOptions);
-}
+//
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:24]: Tests commented

**Review by tkilias — APPROVED** [review]

**Pull request merged**